libcontainer/specconv/spec_linux.go

```diff
@@ -968,6 +969,21 @@ func setupUserNamespace(spec *specs.Spec, config *configs.Config) error {
 		config.UIDMappings = toConfigIDMap(spec.Linux.UIDMappings)
 		config.GIDMappings = toConfigIDMap(spec.Linux.GIDMappings)
 	}
+	if path := config.Namespaces.PathOf(configs.NEWUSER); path != "" {
+		// Cache the current userns mappings in our configuration, so that we
+		// can calculate uid and gid mappings within runc. These mappings are
+		// never used for configuring the container if the path is set.
+		uidMap, gidMap, err := userns.GetUserNamespaceMappings(path)
+		if err != nil {
+			return fmt.Errorf("failed to cache mappings for userns: %w", err)
+		}
+		config.UIDMappings = uidMap
+		config.GIDMappings = gidMap
```

Comment:
If we just need this to be set _after_ the function finishes, can't we defer this? Just defer a function that sets this. And we can just validate things here. Would that work?

Reply:
Anywhere that does `config.HostUID` or `config.HostGID` uses this, which is a fair few places, and spawning a subprocess each time is quite expensive.

However, even without the performance concerns, it needs to be cached here because the alternative is that we have to do `userns.GetUserNamespaceMappings(path)` from within `libcontainer/configs` but because we have to return config-typed objects you need to copy the `GetUserNamespaceMappings` code into the `configs` package.

You can take a look at @lifubang's previous review to see what the issue with that approach is.

> If we just need this to be set after the function finishes, can't we defer this?

Defer as in `defer`, or defer as in `sync.Once`? I don't see how `defer` would change anything -- the code will still run within this function. The purpose of this code is to copy the mappings, not necessarily validation (the userns validation is added in a later commit, and it's added here because we must cache the mappings here).

Reply:
Oh, I missed this is part of `setupUserNamespace()`, yeah, that trick can't work here.